### PR TITLE
WIP: e2e tests

### PR DESCRIPTION
**Pull request by ashetty1**
Few changes still pending.

Pushing this to get some initial feedback on the structure of the test cases.

### CODE REVIEW

File: tests/e2e/e2e_test.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    "fmt"
+    "os/exec"
+    "os"
+    "time"
+    "net/http"
+    "log"
+    "strconv"
+)
+
+var TmpDir = "/tmp/ocdev"
```
Comment (kadel) [inline tests/e2e/e2e_test.go]: it would be better to use ioutil.TmpDir

**Comment by ashetty1** [comment]
@containscafeine @kadel review/feedback please.

**Comment by kadel** [comment]
@ashetty1 please keep vendor updates in separate commit next time. It will make it easier to review

File: tests/e2e/e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	return app
+}
+
+func getCmp() string {
+	cmp := runCmd("ocdev component get --short")
+	return cmp
+}
+
+func pingSvc(url string) {
+	var ep int
```
Comment (kadel) [inline tests/e2e/e2e_test.go]: shouldn't this be just `bool`?

Comment (kadel) [inline tests/e2e/e2e_test.go]: shouldn't this be just `bool`?

File: tests/e2e/e2e_test.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		grepBeforePush := runCmd("curl -s " + pingUrl +
+			" | grep 'Welcome to your Node.js application on OpenShift'")
+
+		log.Printf("Text before change: %s", grepBeforePush)
+
+		// Make changes to the html file
+		runCmd("sed -i 's/Welcome to your Node.js application on OpenShift/Welcome to your Node.js on OCDEV/g' " + TmpDir + "/nodejs-ex/views/index.html")
+
+		// Push the changes
+		runCmd("ocdev component push --dir " + TmpDir + "/nodejs-ex")
```
Comment (kadel) [inline tests/e2e/e2e_test.go]: this command was renamed to just `ocdev push` ( #141)

File: tests/e2e/e2e_test.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	It("create component", func() {
+		runCmd("git clone https://github.com/openshift/nodejs-ex " + TmpDir + "/nodejs-ex")
+		runCmd("ocdev component create nodejs --dir " + TmpDir + "/nodejs-ex")
+		Expect(getCmp()).To(Equal("nodejs"))
+		time.Sleep(10)
+	})
+
+	It("push changes to component", func() {
+
+		// Create route
```
Comment (kadel) [inline tests/e2e/e2e_test.go]: this comment is not correct you are not creating route anywhere

File: tests/e2e/e2e_test.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		getIP := runCmd("oc get svc --no-headers | awk '{ print $2 }' | tr -d '\n'")
```
Comment (kadel) [inline tests/e2e/e2e_test.go]: you can get the whole url with one command without needing `awk` or `tr`.
`oc` has support for go templates
```
oc get svc nodejs -o go-template="{{.spec.clusterIP}}:{{(index .spec.ports 0).port}}"
```

File: tests/e2e/e2e_test.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+func pingSvc(url string) {
+	var ep int
+	pingTimeout := time.After(10 * time.Minute)
+	tick := time.Tick(time.Second)
+
+	for {
+		select {
+		case <-pingTimeout:
+			fmt.Errorf("could not ping the specific service in given time: 10 minutes")
```
Comment (kadel) [inline tests/e2e/e2e_test.go]: `Errorf` without formatting argument. Use just `Error` instead

File: tests/e2e/e2e_test.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+			if err != nil {
+				time.Sleep(1 * time.Second)
+				continue
+			}
+
+			if response.Status == "200 OK" {
+				ep = 1
+
+			} else {
+				fmt.Errorf("for service")
```
Comment (kadel) [inline tests/e2e/e2e_test.go]: Errorf without formatting argument. Use just Error instead

**Comment by kadel** [comment]
Who stable `oc cluster up` is in Travis-CI?

**Comment by kadel** [comment]
I suspect that e2e tests might end up be quite time-consuming.
 We should probably modify Makefile so `make test` just runs uni tests. and create separate `make test-e2e` to run just e2e tests

**Comment by kadel** [comment]
when test fails I don't  see why :-( I just see this:
```
2018/03/05 14:02:46 Error running command ocdev application create usecase5: exit status 1
```

**Comment by kadel** [comment]
Also there is not `oc cluster down` after tests are finihed. After tests are finished you end up with running OpenShift cluster.

**Comment by ashetty1** [comment]
@kadel you think we should manage ocluster up/down from with the test script? Thought it might be best to keep it out of this so that we could run it without modifying on minishift. WDYT?

File: tests/e2e/e2e_test.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+func TestOCdev(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "ocdev test suite")
+}
+
+var _ = BeforeSuite(func() {
+	runCmd("oc cluster down")
+	runCmd("oc cluster up")
```
Comment (ashetty1) [inline tests/e2e/e2e_test.go]: @kadel need you input on this: do you think we should deal with `oc cluster up/down` outside of e2e scripts. One advantage with that would be to allow running tests with minishift.

Comment (kadel) [reply tests/e2e/e2e_test.go]: It might be a good idea. It will also allow easily switch between OpenShift versions

**Comment by ashetty1** [comment]
@containscafeine yes, It would be nice; I thought I get some sort of skeletal structure in and then detailed permutation/combination. WDYT?

**Comment by ashetty1** [comment]
@kadel `component create` using nodejs-ex consistently fails on travis. Tried several times and it times out. Maybe I have to add a different app to get things moving on travis:

```
 ocdev create nodejs --dir /tmp/nodejs-ex
--> Found image 45396e7 (9 days old) in image stream "openshift/nodejs" under tag "6" for "nodejs"
    Node.js 6 
    --------- 
    Node.js 6 available as docker container is a base platform for building and running various Node.js 6 applications and frameworks. Node.js is a platform built on Chrome's JavaScript runtime for easily building fast, scalable network applications. Node.js uses an event-driven, non-blocking I/O model that makes it lightweight and efficient, perfect for data-intensive real-time applications that run across distributed devices.
    Tags: builder, nodejs, nodejs6
    * A source build using binary input will be created
      * The resulting image will be pushed to image stream "nodejs:latest"
      * A binary build was created, use 'start-build --from-dir' to trigger a new build
    * This image will be deployed in deployment config "nodejs"
    * Port 8080/tcp will be load balanced by service "nodejs"
      * Other containers can access this service through the hostname "nodejs"
--> Creating resources with label app=usecase1,app.kubernetes.io/component-name=nodejs,app.kubernetes.io/name=usecase1 ...
    imagestream "nodejs" created
    buildconfig "nodejs" created
    deploymentconfig "nodejs" created
    service "nodejs" created
--> Success
    Build scheduled, use 'oc logs -f bc/nodejs' to track its progress.
    Application is not exposed. You can expose services to the outside world by executing one or more of the commands below:
     'oc expose svc/nodejs' 
    Run 'oc status' to view your app.
please wait, building application...
command: [/usr/bin/oc start-build nodejs --follow --from-dir /tmp/nodejs-ex] failed to run:
Uploading directory "/tmp/nodejs-ex" as binary input for the build ...
Error from server (BadRequest): unable to wait for build nodejs-2 to run: timed out waiting for the condition
: exit status 1
```

**Comment by kadel** [comment]
can you see what is happening with the build?. Is it failing while uploading data to cluster or while building? 
Have you tried it with git? (`ocdev create nodejs --git https://github.com/openshift/nodejs-ex.git`) It might give us some indications what is wrong

**Comment by ashetty1** [comment]
Tracking this in another PR #242. So closing this.

**Pull request closed without merging**